openhands/sdk/agent/agent.py

```diff
@@ -198,63 +189,133 @@ def step(
+                        "LLM provided a security_risk but no security analyzer is "
+                        "configured - THIS SHOULD NOT HAPPEN!"
+                    )
+                try:
+                    security_risk = risk.SecurityRisk(_predicted_risk)
+                except ValueError:
+                    logger.warning(
+                        f"Invalid security_risk value from LLM: {_predicted_risk}"
+                    )
+            action: ActionBase = tool.action_from_arguments(arguments)
+            action_event = ActionEvent(
+                action=action,
+                thought=thought if i == 0 else [],
+                reasoning_content=reasoning_content if i == 0 else None,
+                tool_name=tool.name,
+                tool_call_id=tool_call.id,
+                tool_call=tool_call,
+                llm_response_id=llm_response.id,
+                security_risk=security_risk,
+            )
+            on_event(action_event)
+            action_events.append(action_event)
+
+        # Handle confirmation mode - exit early if actions need confirmation
+        if self._requires_user_confirmation(state, action_events):
+            return
+
+        self._execute_action_events(state, action_events, on_event)
+
+    def _validate_tool_call(
+        self,
+        tool_call: ChatCompletionMessageToolCall,
+    ) -> AgentErrorEvent | None:
+        """Validate a tool call.
+
+        Returns None if validation passes, AgentErrorEvent if validation fails.
+        """
+        # Check if tool call is of type 'function'
+        if tool_call.type != "function":
+            return AgentErrorEvent(
+                error=(
+                    "LLM returned tool calls but some are not of type 'function' - "
+                    "ignoring those"
+                ),
+                tool_name=tool_call.function.name or "unknown",
+                tool_call_id=tool_call.id,
+            )
+
+        # Check if tool exists
+        tool_name = tool_call.function.name
+        if tool_name is None:
+            return AgentErrorEvent(
+                error="Tool call must have a name",
+                tool_name="unknown",
+                tool_call_id=tool_call.id,
+            )
+
+        tool = self.tools_map.get(tool_name, None)
+        if tool is None:
+            available = list(self.tools_map.keys())
+            err = f"Tool '{tool_name}' not found. Available: {available}"
+            return AgentErrorEvent(
+                error=err,
+                tool_name=tool_name,
+                tool_call_id=tool_call.id,
+            )
+
+        # Validate arguments
+        try:
+            arguments = json.loads(tool_call.function.arguments)
+            # Remove security_risk if present (handled separately)
+            arguments.pop("security_risk", None)
+            # Try to create action to validate arguments
+            tool.action_from_arguments(arguments)
```

Comment:
IMO it feels to me that this is making things even more complex -- we do `tool.action_from_arguments` twice now.

Also, right now you have to jump around different place/functions to understand these logics.

I don't think this will meaningful simplify stuff..